### PR TITLE
[SPARK-33298][CORE] Decouple file naming from FileCommitProtocol

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This PR is to propose to decouple file naming functionality from `FileCommitProtocol`. Currently `FileCommitProtocol` mainly does three things:
* commit task output by renaming from stage file name to final file name
* commit job output by renaming directory for dynamic partitions query or writing to custom location
* specify task staging output file name, and final output file name if needed

A `FileCommitProtocol` should cover first two functionalities, but not the 3rd (naming output file). The file commit protocol (by its name and design) should take care of committing output (e.g. rename file and directory), but it should not control what's the file name as well. It should leave caller the flexibility to specify file names (e.g. Hive/Presto/Trino compatible bucket file name is different from what Spark has now - https://github.com/apache/spark/pull/30003). So here the PR is to decouple the file naming from `FileCommitProtocol`.

The changes are:

* Introduce an interface `FileNamingProtocol` to specify how to do output file naming. Add implementation `BatchFileNamingProtocol` for batch queries and `StreamingFileNamingProtocol` for streaming queries.
* Modify existing method `newTaskTempFile` and `newTaskTempFileAbsPath` in `FileCommitProtocol` to allow commit protocol to be notified when a new task output file is added. The input is relative file name/path, and output is the full file path.
* Change `FileFormatDataWriter` to call `FileNamingProtocol` to get relative file path and `FileCommitProtocol` to get full file path.

### Why are the changes needed?

To make commit protocol clearer and allow future flexibility to specify Spark output file name.
Pre-requisite of https://github.com/apache/spark/pull/30003.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit test e.g. `DataFrameReaderWriterSuite.scala` and `InsertSuite.scala`
Will add more unit tests if required. This PR anyway is a code refactoring.

### CODE REVIEW

**Comment by c21** [comment]
cc @cloud-fan could you help take a look when you have time?
Will craft more unit tests if we have consensus on overall design, thanks.

File: core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala [hunk 1]
```diff
@@ -100,9 +101,34 @@ abstract class FileCommitProtocol extends Logging {
...
+   * different tasks will not conflict. This API should be preferred to use instead of deprecated
+   * [[newTaskTempFile]] and [[newTaskTempFileAbsPath]].
+   */
+  def newTaskFile(
+      taskContext: TaskAttemptContext,
+      stagingPath: String,
+      finalPath: Option[String],
+      stagingDir: Option[String]): Unit = {
+    // No-op as default implementation to be backward compatible with custom [[FileCommitProtocol]]
+    // implementations before Spark 3.2.0.
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala:129]: I'd like to understand more about backward compatibility. If there is an old impl of file commit protocol, in Spark 3.2 the impl will commit nothing, and this silent "noop" is even worse than an explicit compile error.

This is a developer API, if there is really no way to keep backward compatibility, let's just break the API and force people to correct their impl according to the new API.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1192,6 +1192,13 @@ object SQLConf {
       .createWithDefault(
         "org.apache.spark.sql.execution.datasources.SQLHadoopMapReduceCommitProtocol")
 
+  val FILE_NAMING_PROTOCOL_CLASS =
+    buildConf("spark.sql.sources.namingProtocolClass")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1196]: do we really need to allow users to customize the file naming?

File: core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala (hunk 1 above)
Comment (c21) [reply core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala]: @cloud-fan - I think it's impossible for me to keep backward compatibility cleanly. Updated here to just introduce new method and break the API.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1196]: @cloud-fan - yes I think so. Before this PR, file naming is part of commit protocol, and each commit protocol checked in Spark code base, has their own specification for naming - `HadoopMapReduceCommitProtocol`, `PathOutputCommitProtocol` and `ManifestFileCommitProtocol`. So the external commit protocol we should expect they might already have their custom way of naming, and we should allow them to implement their own naming protocol.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1196]: I agree that the file naming shouldn't be part of commit protocol, but I think the callers in Spark should decide the file name, like file format writer, or something similar for hive tables, not users.

**Comment by c21** [comment]
@cloud-fan - thank you for offline discussion and update the PR to use the discussed approach. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Updated per discussion offline. Keep naming protocol internal and not user-facing via config.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -1405,9 +1405,9 @@ object QueryExecutionErrors {
       s"Multiple streaming queries are concurrently using $path", e)
   }
 
-  def addFilesWithAbsolutePathUnsupportedError(commitProtocol: String): Throwable = {
+  def addFilesWithAbsolutePathUnsupportedError(protocol: String): Throwable = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1408]: why this change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 4]
```diff
@@ -96,7 +96,7 @@ object FileFormatWriter extends Logging {
       sparkSession: SparkSession,
       plan: SparkPlan,
       fileFormat: FileFormat,
-      committer: FileCommitProtocol,
+      protocols: (FileCommitProtocol, FileNamingProtocol),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:99]: why not pass 2 parameter?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 3 above)
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1408]: @cloud-fan - sorry, this does not need. Was calling it from naming protocol as well in one iteration. Will change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 4 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:99]: From my side, whenever I add the 11th parameter of a scala method, intellij will mark it as a lint error. Do we have # of parameter rule in Spark?

File: core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ * The context for Spark output file. This is used by [[FileNamingProtocol]] to create file path.
+ *
+ * @param ext Source specific file extension, e.g. ".snappy.parquet".
+ * @param relativeDir Relative directory of file. Can be used for writing dynamic partitions.
+ *                    E.g., "a=1/b=2" is directory for partition (a=1, b=2).
+ * @param prefix file prefix.
+ */
+final case class FileContext(
+  ext: String,
+  relativeDir: Option[String],
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: I'm not very sure about it. It seems more clear to me to let `FileNamingProtocol` only generate filename, and the caller side should construct the proper relative path w.r.t. the partition dir.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingFileNamingProtocol.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  override def getTaskTempPath(
+      taskContext: TaskAttemptContext, fileContext: FileContext): String = {
+    // The file name looks like part-r-00000-2dd664f9-d2c4-4ffe-878f-c6c70c1fb0cb_00003.gz.parquet
+    // Note that %05d does not truncate the split number, so if we have more than 100000 tasks,
+    // the file name is fine and won't overflow.
+    val split = taskContext.getTaskAttemptID.getTaskID.getId
+    val uuid = UUID.randomUUID.toString
+    val ext = fileContext.ext
+    val filename = f"part-$split%05d-$uuid$ext"
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingFileNamingProtocol.scala:40]: shall we fail here if `prefix` is not None? or support `prefix` here?

File: core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala (hunk 5 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: same to `ext` and `prefix`: can we let the caller side to pretend/append `ext`/`prefix` to the generated file name?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: All the information in `FileContext` is not something that the impl can customize: the generated file name must have `ext` and the end, `prefix` at the beginning, and `relativeDir` as the parent dir. Then it's better to let the caller side to guarantee it.

File: core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala [hunk 7]
```diff
@@ -73,35 +73,31 @@ abstract class FileCommitProtocol extends Logging {
...
+   * the file will be at the location specified by the "finalPath" after job commits.
+   *
+   * Important: it is the caller's responsibility to add uniquely identifying content to
+   * "relativePath" and "finalPath" if a task is going to write out multiple files to the same
+   * directory. The file commit protocol only guarantees that files written by different tasks will
+   * not conflict.
    */
   def newTaskTempFileAbsPath(
-      taskContext: TaskAttemptContext, absoluteDir: String, ext: String): String
+      taskContext: TaskAttemptContext, relativePath: String, finalPath: String): String
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala:100]: note: I can see that, this API makes the code simpler, but it makes the semantic a bit more complicated: what if the final path doesn't have the same name as the `relativePath`? Maybe it's better to have `fileName: String, finalDir: String`. Then the semantic is pretty here: the impl should commit the new file to the target dir.

File: core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala (hunk 5 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: nvm, `FileNamingProtocol` needs to return an absolute path, so we must provide all these info as a `FileContext` and ask the impl to follow.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: And the API can be simply named `getTaskFileName`.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: This API is only to abstract the naming differences between batch and streaming file writing. If that's not necessary, maybe we can remove this abstraction entirely.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: `jobId` is generated with UUID as well, I don't see why streaming write needs to generate a new UUID per file, instead of using job id.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/io/FileNamingProtocol.scala:56]: I'm cleaning this up in https://github.com/apache/spark/pull/33002

**Comment by c21** [comment]
Update: we decided to go with https://github.com/apache/spark/pull/33012 instead of this PR, as we know [some other projects](https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/files/DelayedCommitProtocol.scala#L116) are using the existing APIs to customize file name,  so we don't want to break them for now.
